### PR TITLE
feat: Add ImageList class

### LINKED ISSUE #528: Add a new ImageSet Class
### Is your feature request related to a problem?

The user should be able to combine more than one image together. 

### Desired solution

Add a ImageSet class.
This new class should deliver the same functionality as the Image class and furthermore have methods for batching for the training of NN models.

### Possible alternatives (optional)

_No response_

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by Marsmaennchen221**
Closes #528, #599, #600

### Summary of Changes

Added ImageList class.

### CODE REVIEW

**Comment by Marsmaennchen221** [comment]
@WinPlay02 Please look over the hash and size implementation

**Review by WinPlay02 — APPROVED** [review]
Implementations of hash and sizeof look good, as all available information is included without hashing the complete tensor (untested though)

**Comment by lars-reimann** [comment]
@Safe-DS/library: Please review each other's code. Always waiting for me delays thing unnecessarily.

File: src/safeds/data/image/containers/_multi_size_image_list.py [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+            image_list._image_list_dict[size] = _SingleSizeImageList._create_image_list([image_tensor], [index])
+
+        if image_tensor.size(dim=0) != self.channel:
+            image_list = image_list.change_channel(
+                max(image_tensor.size(dim=0), self.channel),
+            )._as_multi_size_image_list()
+
+        return image_list
+
+    def add_images(self, images: list[Image] | ImageList) -> ImageList:
```
Comment (Gerhardsa0) [inline src/safeds/data/image/containers/_multi_size_image_list.py:271]: shouldn't the return type be multi_image_list?
for the most functions in the class

File: src/safeds/data/image/containers/_single_size_image_list.py [hunk 2]
```diff
@@ -0,0 +1,720 @@
...
+            else:  # image channel 3 and max channel 4
+                image_to_append = torch.cat([image, torch.full(image[0:1].size(), 255, device=image.device)], dim=0)
+            images_ready_to_concat.append((image_to_append.unsqueeze(dim=0), index))
+        image_list._tensor = torch.cat([image for image, index in images_ready_to_concat])
+        image_list._tensor_positions_to_indices = [index for image, index in images_ready_to_concat]
+        image_list._indices_to_tensor_positions = image_list._calc_new_indices_to_tensor_positions()
+
+        return image_list
+
+    def clone(self) -> ImageList:
```
Comment (Gerhardsa0) [inline src/safeds/data/image/containers/_single_size_image_list.py:92]: same here with the single size image list?

File: src/safeds/data/image/containers/_multi_size_image_list.py [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+    | [from_images][safeds.data.image.containers._image_list.ImageList.from_images] | Create an ImageList from a list of Images.               |
+    | [from_files][safeds.data.image.containers._image_list.ImageList.from_files]   | Create an ImageList from a directory or a list of files. |
+    """
+
+    def __init__(self) -> None:
+        self._image_list_dict: dict[tuple[int, int], ImageList] = {}  # {image_size: image_list}
+        self._indices_to_image_size_dict: dict[int, tuple[int, int]] = {}  # {index: image_size}
+
+    @staticmethod
+    def _create_image_list(images: list[Tensor], indices: list[int]) -> ImageList:
```
[the remark this answers is not in the document]
Comment (lars-reimann) [reply src/safeds/data/image/containers/_multi_size_image_list.py:48]: In this case it's actually desireable to only expose the "interface" `ImageList`. You can do the same operations on it, so there's no reason to expose the concrete implementation.

File: src/safeds/data/image/containers/_image_list.py [hunk 4]
```diff
@@ -0,0 +1,928 @@
...
+            The image to be removed from the image list
+
+        Returns
+        -------
+        image_list:
+            the image list with the given image removed
+        """
+        return self.remove_image_by_index(self.index(image))
+
+    def remove_images(self, images: list[Image]) -> ImageList:
```
Comment (Gerhardsa0) [inline src/safeds/data/image/containers/_image_list.py:558]: does this raise an error?
Something like Indexoutofbounds

File: src/safeds/data/image/containers/_image_list.py [hunk 5]
```diff
@@ -0,0 +1,928 @@
...
+        indices = list(range(len(images)))
+        first_width = images[0].width
+        first_height = images[0].height
+        for im in images:
+            if first_width != im.width or first_height != im.height:
+                return _MultiSizeImageList._create_image_list([image._image_tensor for image in images], indices)
+        return _SingleSizeImageList._create_image_list([image._image_tensor for image in images], indices)
+
+    @staticmethod
+    def from_files(path: str | Path | Sequence[str | Path]) -> ImageList:
```
[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image_list.py:89]: `pil_image_open` already raises an error but I will add this to the docstrings and a test

File: src/safeds/data/image/containers/_image_list.py [hunk 6]
```diff
@@ -0,0 +1,928 @@
...
+        if isinstance(self, _SingleSizeImageList):
+            return self
+        raise ValueError("The given image_list is not a SingleSizeImageList")
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # Transformations
+    # ------------------------------------------------------------------------------------------------------------------
+
+    @abstractmethod
+    def change_channel(self, channel: int) -> ImageList:
```
[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image_list.py:467]: added

File: tests/safeds/data/image/containers/test_image_list.py [hunk 7]
```diff
@@ -0,0 +1,1249 @@
...
+
+
+class TestFromFiles:
+
+    @pytest.mark.parametrize(
+        "resource_path",
+        [images_all(), str(test_images_folder), *images_all(), [Path(im) for im in images_all()], test_images_folder, *[Path(im) for im in images_all()]],
+        ids=["all-images", "images_folder", *images_all_ids(), "all-images-path", "images_folder-path", *[s + "-path" for s in images_all_ids()]],
+    )
+    def test_from_files_creation(self, resource_path: str | Path | Sequence[str | Path], snapshot_png_image_list: SnapshotAssertion) -> None:
```
Comment (Marsmaennchen221) [inline tests/safeds/data/image/containers/test_image_list.py:401]:
```suggestion
    def test_from_files_creation(self, resource_path: str | Path | list[str | Path], snapshot_png_image_list: SnapshotAssertion) -> None:
```

File: tests/safeds/data/image/containers/test_image_list.py [hunk 8]
```diff
@@ -0,0 +1,1249 @@
...
+        torch.set_default_device(torch.device("cpu"))
+        image_list = ImageList.from_files(resolve_resource_path(resource_path))
+        assert image_list == snapshot_png_image_list
+
+    @pytest.mark.parametrize(
+        "resource_path",
+        ["\\images\\missing_directory\\", Path("\\images\\missing_directory\\"), ["\\images\\missing_file1.png", "\\images\\missing_file2.png"], [Path("\\images\\missing_file1.png"), Path("\\images\\missing_file2.png")], [*images_all(), "\\images\\missing_file2.png"], [*[Path(im) for im in images_all()], Path("\\images\\missing_file2.png")]],
+        ids=["dir-str", "dir-path", "list-str", "list-path", "list-str-last-missing", "list-path-last-missing"]
+    )
+    def test_should_raise_if_one_file_or_directory_not_found(self, resource_path: str | Path | Sequence[str | Path]) -> None:
```
Comment (Marsmaennchen221) [inline tests/safeds/data/image/containers/test_image_list.py:411]:
```suggestion
    def test_should_raise_if_one_file_or_directory_not_found(self, resource_path: str | Path | list[str | Path]) -> None:
```

File: tests/safeds/data/image/containers/test_image_list.py [hunk 9]
```diff
@@ -0,0 +1,1249 @@
+import random
+import sys
+import tempfile
+from pathlib import Path
+from typing import Sequence
```
Comment (Marsmaennchen221) [inline tests/safeds/data/image/containers/test_image_list.py:5]:
```suggestion
```

File: tests/safeds/data/image/containers/test_image_list.py [hunk 10]
```diff
@@ -0,0 +1,1248 @@
...
+        torch.set_default_device(torch.device("cpu"))
+        image_list = ImageList.from_files(resolve_resource_path(resource_path))
+        assert image_list == snapshot_png_image_list
+
+    @pytest.mark.parametrize(
+        "resource_path",
+        ["\\images\\missing_directory\\", Path("\\images\\missing_directory\\"), ["\\images\\missing_file1.png", "\\images\\missing_file2.png"], [Path("\\images\\missing_file1.png"), Path("\\images\\missing_file2.png")], [*images_all(), "\\images\\missing_file2.png"], [*[Path(im) for im in images_all()], Path("\\images\\missing_file2.png")]],
+        ids=["dir-str", "dir-path", "list-str", "list-path", "list-str-last-missing", "list-path-last-missing"]
+    )
+    def test_should_raise_if_one_file_or_directory_not_found(self, resource_path: str | Path | list[str | Path]) -> None:
```
Comment (Marsmaennchen221) [inline tests/safeds/data/image/containers/test_image_list.py:410]:
```suggestion
    def test_should_raise_if_one_file_or_directory_not_found(self, resource_path: str | Path) -> None:
```

**Comment by WinPlay02** [comment]
@Marsmaennchen221 See #617 for changes to the hashing (+ merging master)

File: src/safeds/data/image/containers/_image_list.py [hunk 11]
```diff
@@ -0,0 +1,983 @@
...
+
+    @abstractmethod
+    def to_jpeg_files(self, path: str | Path | list[str | Path]) -> None:
+        """
+        Save all images as jpeg files.
+
+        Parameters
+        ----------
+        path:
+            Either the path to a directory or a list of directories which has directories for either all different sizes or all different images
```
Comment (Marsmaennchen221) [inline src/safeds/data/image/containers/_image_list.py:381]:
```suggestion
            Either the path to a directory or a list of directories which has directories for either all different sizes or all different images. Any non-existant path will be created
```

File: src/safeds/data/image/containers/_image_list.py [hunk 12]
```diff
@@ -0,0 +1,983 @@
...
+
+        Raises
+        ------
+        IllegalFormatError
+            If the channel of the images is not supported
+        ValueError
+            If the path is a list but has too few or too many entries
```
[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image_list.py:388]: No, the given path will be created in this case. I updated the docstring to include this behavior

File: src/safeds/data/image/containers/_image_list.py [hunk 13]
```diff
@@ -0,0 +1,983 @@
...
+        Raises
+        ------
+        ValueError
+            If the path is a list but has too few or too many entries
+        """
+
+    @abstractmethod
+    def to_images(self, indices: list[int] | None = None) -> list[Image]:
+        """
+        Return a list of all images in this image list.
```
[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/image/containers/_image_list.py:410]: If indices is `None`, all indices will be included in the output. I updated the docstring to include this behavior

File: src/safeds/data/image/containers/_image_list.py [hunk 14]
```diff
@@ -0,0 +1,983 @@
...
+
+    @abstractmethod
+    def to_png_files(self, path: str | Path | list[str | Path]) -> None:
+        """
+        Save all images as png files.
+
+        Parameters
+        ----------
+        path:
+            Either the path to a directory or a list of directories which has directories for either all different sizes or all different images
```
Comment (Marsmaennchen221) [inline src/safeds/data/image/containers/_image_list.py:399]:
```suggestion
            Either the path to a directory or a list of directories which has directories for either all different sizes or all different images. Any non-existant path will be created
```

**Review by Gerhardsa0 — APPROVED** [review]

**Pull request merged**